### PR TITLE
only apply error models to UDP packets

### PR DESCRIPTION
**Pull request by marten-seemann**
cc @larseggert 

This was correct all the time, but apparently the addition of IPv6 leads to the frequent sending of ICMPv6 packets (Neighbor Soliciation messages).

### CODE REVIEW

File: sim/scenarios/corrupt-rate/corrupt-rate-error-model.cc [hunk 1]
```diff
@@ -30,6 +30,8 @@ CorruptRateErrorModel::CorruptRateErrorModel()
 void CorruptRateErrorModel::DoReset(void) { }
 
 bool CorruptRateErrorModel::DoCorrupt(Ptr<Packet> p) {
+    if(!IsUDPPacket(p)) return false;
```
Comment (janaiyengar) [inline sim/scenarios/corrupt-rate/corrupt-rate-error-model.cc:33]: Add a comment here saying that this applies only to PPP links, since IsUDPPacket assumes it.

File: sim/scenarios/drop-rate/drop-rate-error-model.cc [hunk 2]
```diff
@@ -22,6 +22,8 @@ DropRateErrorModel::DropRateErrorModel()
 void DropRateErrorModel::DoReset(void) { }
  
 bool DropRateErrorModel::DoCorrupt(Ptr<Packet> p) {
+    if(!IsUDPPacket(p)) return false;
```
Comment (janaiyengar) [inline sim/scenarios/drop-rate/drop-rate-error-model.cc:25]: same as above

File: sim/scenarios/helper/quic-packet.cc [hunk 3]
```diff
@@ -6,17 +6,48 @@
...
+#include "ns3/ipv6-header.h"
 #include "ns3/udp-header.h"
 
 using namespace ns3;
 using namespace std;
 
+
+bool IsUDPPacket(Ptr<Packet> p) {
+    PppHeader ppp_hdr = PppHeader();
+    p->RemoveHeader(ppp_hdr);
```
Comment (janaiyengar) [inline sim/scenarios/helper/quic-packet.cc:18]: Maybe check if PPP header and return early with error if it isn't.

File: sim/scenarios/droplist/droplist-error-model.cc [hunk 4]
```diff
@@ -18,9 +19,12 @@ DroplistErrorModel::DroplistErrorModel()
...
  
 bool DroplistErrorModel::DoCorrupt(Ptr<Packet> p) {
-    if(drops.find(++packet_num) == drops.end())
-        return false;
-    cout << "Dropping packet number " << packet_num << endl;
+    if(!IsUDPPacket(p)) return false;
+    if(drops.find(++packet_num) == drops.end()) return false;
+    
+    QuicPacket qp = QuicPacket(p);
+    cout << "Dropping packet " << packet_num << " (" << qp.GetUdpPayload().size() << " bytes) from " << qp.GetIpv4Header().GetSource() << endl;
```
Comment (janaiyengar) [inline sim/scenarios/droplist/droplist-error-model.cc:26]: This introduces a requirement for the packet to be an IPv4 packet, right? Why is this needed?

**Review by janaiyengar — COMMENTED** [review]
A few comments

File: sim/scenarios/helper/quic-packet.cc (hunk 3 above)
Comment (marten-seemann) [reply sim/scenarios/helper/quic-packet.cc:18]: The PPP header here is just 2 bytes indicating the protocol. Not sure how I would test for it.

File: sim/scenarios/droplist/droplist-error-model.cc (hunk 4 above)
Comment (marten-seemann) [reply sim/scenarios/droplist/droplist-error-model.cc:26]: That's a general requirement of `QuicPacket` at the moment (which is why I added the TODO). We'll need a general fix for that (any suggestions for the class API?). Just added it here to get some more useful logging here.

Comment (janaiyengar) [reply sim/scenarios/droplist/droplist-error-model.cc:26]: Understood. I would suggest that QuicPacket can be a class that handles both v4 and v6 packets; a "QUIC packet" ought to be independent of IP version.

One way to do this might be to have the QuicPacket constructor do the same check you do now in IsUDPPacket() and store v4 or v6 address. You can then have QuicPacket.GetIpv4Address() and QuicPacket.GetIpv6Address() methods that either return empty string or the address, and a QuicPacket.GetIpAddress() method that returns whichever one is populated.

This doesn't have to be done in this PR, but given that we're trying to support v6, it might be a good idea to not introduce more v4-only dependencies.

**Review by janaiyengar — APPROVED** [review]
LGTM, modulo the nits I've noted.
